packages/cli/test/cases/develop.ssr/src/pages/post.js

```diff
@@ -0,0 +1,26 @@
+export default class PostPage extends HTMLElement {
+  constructor({ post = {} }) {
+    super();
+    this.post = post;
+  }
+
+  async connectedCallback() {
+    const { id, title, body } = this.post;
+
+    this.innerHTML = `
+      <h1>Fetched Post ID: ${id}</h1>
+      <h2>${title}</h2>
+      <p>${body}</p>
+    `;
+  }
+}
+
+export async function loader(request) {
```

Comment:
Another thought here is to just lean into the `class` based paradigm instead of creating a custom API, e.g. `loaders`.  So basically something like this, wherein the `request` comes through the `constructor` instead, any synchronous operations are done there, and let all the `async` work happen in `connectedCallback`
```js
export default class PostPage extends HTMLElement {
  constructor(request) {
    super();
    
    const params = new URLSearchParams(request.url.slice(request.url.indexOf('?')));
    this.postId = params.get('id');
  }

  async connectedCallback() {
    const { postId } = this;
    const post = await fetch(`https://jsonplaceholder.typicode.com/posts/${postId}`).then(resp => resp.json());
    const { id, title, body } = post;

    this.innerHTML = `
      <h1>Fetched Post ID: ${id}</h1>
      <h2>${title}</h2>
      <p>${body}</p>
    `;
  }
}
```

Reply:
Per the last Greenwood weekly meeting, a good suggestion was made that the default behavior could be done as in the snippet above, and that possibly for #880 , for users that want to _opt-in_ to hydration, that mechansim could be to export a `loader` function.

This would have a few benefits:
- It helps draw a line between a basic component definition, and a custom bespoke API
- For hydration, we need to know the props.  Having a `loader` function allows us to serialize the props for further hydration at runtime.
- Perhaps to tie it all in together, we can leverage #1005 ?